### PR TITLE
Adding mongodb for violations and using oracle for authentication 

### PR DESCRIPTION
**Pull request by ak-seyam**
please review the diff and let me know if something need to be changed

### CODE REVIEW

File: model/user.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+from sqlalchemy import Column, String, Integer
+Base = declarative_base()
+
+class user(Base):
+    __tablename__ = 'USERS'
+    id = Column(Integer,primary_key=True,
+                unique=True,
+                autoincrement=True)
+    car_id = Column(String)
+    user_id = Column(String)
```
Comment (abdo1819) [inline model/user.py]: those cannot be null 
add nullable=False

https://docs.sqlalchemy.org/en/13/core/metadata.html?highlight=column#sqlalchemy.schema.Column.params.nullable

File: model/user.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    id = Column(Integer,primary_key=True,
+                unique=True,
+                autoincrement=True)
+    car_id = Column(String)
+    user_id = Column(String)                                   
+    @property
+    def serialize(self):
+        return {
+            'car_id' : self.car_id,
+            'user_id' : self.time
```
Comment (abdo1819) [inline model/user.py]: self.time ? why 
'user_id' : self.user_id

File: model/user.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy import Column, String, Integer
+Base = declarative_base()
+
+class user(Base):
+    __tablename__ = 'USERS'
+    id = Column(Integer,primary_key=True,
```
Comment (abdo1819) [inline model/user.py]: this table is weird what it represent ?

### if it is table for users __which it seem to be__
> then this is the user id why adding it again as user_id

### if it is relation table between user and car tables
> then this relation would be one user to many cars implemented like here 
https://docs.sqlalchemy.org/en/13/orm/basic_relationships.html#many-to-one
and you need to add cars table

File: model/user.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy import Column, String, Integer
+Base = declarative_base()
+
+class user(Base):
```
Comment (abdo1819) [inline model/user.py]: use CapWords convention for class names   so :     user --> User
https://www.python.org/dev/peps/pep-0008/#class-names

File: names.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+test_database_name = "test3"
+database_name = "evil_car"
+roads = "roads"
+violations_collection = "violations"
+testingCollection = "testingCollection"
+starting_longitude = "startingLongitude"
+starting_latitude = "startingLatitude"
+obstacles_collection = "obstacles"
```
Comment (abdo1819) [inline names.py:8]: what is the use of those ?

they same to be fields names in mongodb document  or collections names
Am i write? if so
can you write the structure of every document as comment here

or better than commenting
i suggest using dict for those names so obstacles document for example would be
```python
obstacles_collection = {name = "obstacles",
                                      long: 'long',
                                      lat: 'lat',
                                      time: 'time'}
```
then we can keep track of both the document structure and its fields names as well collection name

File: db_setup.py [hunk 6]
```diff
@@ -1,36 +1,47 @@
...
+                 db_name="evil_car"):
+        self.db = _create_database(URI,db_name)
+
+    def _get_collection(self,collection_name):
+        return self.db[collection_name]
+
+    def add_data(self,collection_name,data):
+        """
+        """
+        # TODO: add data checkers here
```
Comment (abdo1819) [inline db_setup.py:29]: for the checkers look at my comment at name.py
if u used dictionary you check its fields easly

File: db_setup.py [hunk 7]
```diff
@@ -1,36 +1,47 @@
...
-
-
-Base.metadata.create_all(engine)
+    def get_data(self,collection_name,collection_filter=None):
+        """
+        """
+        col = self._get_collection(collection_name)
+        if collection_filter is not None:
+            return col.find(collection_filter,{"_id":0}) 
+        return col.find({},{"_id":0})
```
Comment (abdo1819) [inline db_setup.py:38]: why exclude the id 
won't we need it if wanna to edit or something ?

idk shouldn't u remove it when sending data not from here

**Review by abdo1819 — COMMENTED** [review]
- remove __pycache__ files  and it folder to ignore

File: names.py (hunk 5 above)
Comment (ak-seyam) [reply names.py:8]: this a was a **lazy ,string-less, coupled** way to make a globally used string because, you know, no one loves string typos, though there might be a better way to do so, i don't know. 
Using dicts will violate the string-less rule and i want to avoid string whenever possible.

File: db_setup.py (hunk 6 above)
Comment (ak-seyam) [reply db_setup.py:29]: i forgot to remove this TODO though, checkers are not appropriate here because add data should do one thing **adding data**.

File: model/user.py (hunk 3 above)
Comment (A-Siam) [reply model/user.py]: >if it is relation table between user and car tables
yes, it is. i'm fixing it now

File: db_setup.py (hunk 7 above)
Comment (ak-seyam) [reply db_setup.py:38]: it was made for visualization purposes, i think will need it only if we are going to provide any deleting mechanism because (in my opinion) violations and obstacles data shouldn't be edited
but this doesn't mean to completely ignore the id and i will try to make add that feature

File: model/user.py (hunk 3 above)
Comment (ak-seyam) [reply model/user.py]: >if it is relation table between user and car tables

tl;dr: it's the middle table between 2 tables we already don't have
it is indeed a relationship between two columns we don't have yet so it's the middle table between the car owner/s and the car/s  
NOTE: i don't know if it's possible for 2 persons to have the same car legally or not but let's keep it as elastic as possible

Comment (ak-seyam) [reply model/user.py]: if it's not it's going to be a simple **unique** constraint on user_id

Comment (A-Siam) [reply model/user.py]: note it's the middle table between 2 tables we already don't have

Comment (abdo1819) [reply model/user.py]: ok

File: names.py (hunk 5 above)
Comment (abdo1819) [reply names.py:8]: but i still think we need a better way to hold those values (cllection-names and fields-key) 
what about class with static properties
```python
from enum import Enum
class obstacles_collection(Enum) 
                                      @property 
                                      def name():
                                             return "obstacles
                                      
                                      @property 
                                      def long():
                                             return "long"

.
. 
.
. so on

 ```
so for example send object function would be
```python
from names import obstacles_collection
def send_obstacles():
 data_manager.add_data(obstacles_collection.name,
                                        {obstacles_collection.long: request.args.get('long'),
                                         obstacles_collection.lat: request.args.get('lat'),
                                         obstacles_collection.time: str(datetime.datetime.now())})
```

instead of 
```python
def send_obstacles():
 data_manager.add_data(names.obstacles_collection, {'long': request.args.get('long'),
                                                       'lat': request.args.get('lat'),
                                                       'time': str(datetime.datetime.now())})
```

or for better practice I suggest using odm for monogdb 
https://github.com/mongodb/pymodm

may add it in another pull request
i will add this as an issue __suggestion__

Comment (abdo1819) [reply names.py:8]: https://github.com/abdo1819/server_evilcar/issues/4

File: db_setup.py (hunk 7 above)
Comment (abdo1819) [reply db_setup.py:38]: > violations and obstacles data shouldn't be edited

we catch violation automatically a human reviewer would be essential to handle grievance from drivers
